R/pairwise-comparisons.R

```diff
@@ -0,0 +1,594 @@
+
+  # mirror computations
+  combinations_mirrored <- data.table::copy(combinations)
+  data.table::setnames(combinations_mirrored,
+                       old = c("model", "compare_against"),
+                       new = c("compare_against", "model"))
+  combinations_mirrored[, ratio := 1/ratio]
+
+  # add a one for those that are the same
+  combinations_equal <- data.table::data.table(model = models,
+                                               compare_against = models,
+                                               ratio = 1,
+                                               pval = 1,
+                                               adj_pval = 1)
+
+  result <- data.table::rbindlist(list(combinations,
+                                       combinations_mirrored,
+                                       combinations_equal),
+                                  use.names = TRUE)
+
+  # make result character instead of factor
+  result[, `:=`("model" = as.character(model),
+                "compare_against" = as.character(compare_against))]
+
+
+  # calculate relative wis as geometric mean
+  # need a different name for the variable! mean_ratio?
+  # small theta is again better. If a baseline is given, exclude it
+  # from the computation of the geometric mean
+  # maybe there is a more elegant way to do this
+  if (!is.null(baseline)) {
+    result_without_baseline <- data.table::copy(result)
+    # filter out all ratios where compare_against is the baseline
+    result_without_baseline <- result_without_baseline[compare_against != baseline, ]
+    result_without_baseline[, `:=` (theta = geom_mean_helper(ratio)),
+                            by = "model"]
+    # merge back to retain the ratios even for comparisons with the baseline
+    result <- merge(result, result_without_baseline, all.x = TRUE)
+    # remove NAs form merge in the thetas
+    result[, theta := unique(na.omit(theta)), by = "model"]
+  } else {
+    result[, `:=` (theta = geom_mean_helper(ratio),
+                   rel_to_baseline = NA_real_),
+           by = "model"]
+  }
+
+  if(!is.null(baseline)) {
+    baseline_theta <- unique(result[model == baseline, ]$theta)
+    result[, rel_to_baseline := theta / baseline_theta]
+  }
+
+  # remove all the rows that are not present in summarise_by before merging
+  cols_to_keep <- unique(c(summarise_by, "model"))
+  cols_to_remove <- colnames(scores)[!(colnames(scores) %in% cols_to_keep)]
+  scores[, eval(cols_to_remove) := NULL]
+  scores <- unique(scores)
+  # allow.cartesian needs to be set as sometimes rows will be duplicated a lot
+  out <- merge(scores, result, by = "model", all = TRUE)
+
+  # rename ratio to mean_scores_ratio
+  data.table::setnames(out, old = c("ratio", "theta", "rel_to_baseline"),
+                       new = c("mean_scores_ratio", "relative_skill", "scaled_rel_skill"))
+
+  return(out)
+}
+
+
+
+
+
+
+
+
+
+
+
+# function to compare two models based on their overlap
```

Comment:
at some point it would be nice to have some documentation for these internal functions. I find the layout of which function calls which here pretty confusing.